Rewrite command and overload builders using C#7 features

FiniteReality opened the pull request:
Passes simple test cases and some more complicated ones, I would appreciate more testers though.

EDIT: I forgot to mention that this is *definitely* a breaking change.

FiniteReality committed:
Rewrite command builders using C#7 features

Not completely tested as of yet

Joe4evr commented on src/Discord.Net.Commands/Builders/CommandBuilder.cs:
```diff
@@ -8,43 +8,41 @@ namespace Discord.Commands.Builders
         }
         //User-defined
-        internal CommandBuilder(ModuleBuilder module, string primaryAlias, Func<ICommandContext, object[], IServiceProvider, Task> callback)
+        internal CommandBuilder(ModuleBuilder module, string primaryAlias, Action<OverloadBuilder> defaultOverloadBuilder)
```
Why go from a `Func<..., Task>` to an `Action<...>`? Just thinking out loud.

FiniteReality replied:
The overload builder contains the callback, so it'd be something like this:

```
module.AddCommand("name", overload => {
    overload.WithCallback(/*callback here*/);
});
```

Auralytical commented on src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs:
```diff
@@ -196,23 +229,29 @@ private static void BuildParameter(ParameterBuilder builder, System.Reflection.P
             foreach (var attribute in attributes)
             {
                 // TODO: C#7 type switch
```
You left a TODO in

FiniteReality replied:
Whoops. At least it was a comment and not missing code :stuck_out_tongue:

FiniteReality committed:
Remove extraneous TODO comment

FiniteReality committed:
Miscellaneous improvements to commands

- OverloadInfo in Before/AfterExecute
  Now you know *exactly* what command is being executed.
- TypeReaders are given the IServiceProvider
  Now you can write TypeReaders for DB-backed data

FiniteReality committed:
Add some debug logging into ModuleClassBuilder

Fixes #675 by logging a message when a module is not public and contains
commands
Also logs how many modules have been successfully built

FiniteReality commented:
After a discussion, we've decided it would be better to find another way to do this.
